rust_snuba/src/types.rs

```diff
@@ -7,26 +7,116 @@ use std::collections::BTreeMap;
 
 pub type CommitLogOffsets = BTreeMap<u16, (u64, DateTime<Utc>)>;
 
-#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+#[derive(Debug, Default, Clone)]
+struct LatencyRecorder {
+    sum_secs: f64,
+    max_secs: i64,
+    num_values: usize,
+}
+
+impl From<DateTime<Utc>> for LatencyRecorder {
+    fn from(value: DateTime<Utc>) -> Self {
+        let value = value.timestamp();
+        LatencyRecorder {
+            sum_secs: value as f64,
+            max_secs: value,
+            num_values: 1,
+        }
+    }
+}
+
+impl LatencyRecorder {
+    fn merge(&mut self, other: Self) {
+        self.sum_secs += other.sum_secs;
+        self.max_secs = max(self.max_secs, other.max_secs);
+        self.num_values += other.num_values;
+    }
+
+    fn send_metric(&self, metrics: &BoxMetrics, write_time: DateTime<Utc>, metric_name: &str) {
+        if self.num_values == 0 {
+            return;
+        }
+
+        let into_latency = |ts: DateTime<Utc>| (write_time - ts).num_seconds().try_into().ok();
+
+        if let Some(ts) = DateTime::from_timestamp(self.max_secs, 0).and_then(into_latency) {
```

Comment:
converting to `DateTime`, just to take the difference as seconds (is that `f64`?) is a bit weird. Why not convert `write_time` to seconds first and then just do straight up math?

Also, do you really want to explicitly log overflows here, or just do saturating math?

Reply:
> is that f64?

yes, i need a type where the sum of multiple timestamps will fit. u64 will overflow easily, and in theory with sums and averages I don't really need that much precision. i could also use u128 i believe

i'll rewrite the function, i agree with your other points